token-swap/program/src/curve/constant_price.rs

```diff
@@ -478,4 +505,58 @@ mod tests {
             assert!(new_value * pool_token_supply >= value * new_pool_token_supply);
         }
     }
+
+    proptest! {
+        #[test]
+        fn curve_value_does_not_decrease_from_withdraw(
+            (pool_token_supply, pool_token_amount) in total_and_intermediate(),
+            swap_token_a_amount in 1..u64::MAX,
+            swap_token_b_amount in 1..u32::MAX, // kept small to avoid proptest rejections
+            token_b_price in 1..u32::MAX, // kept small to avoid proptest rejections
+        ) {
+            let curve = ConstantPriceCurve { token_b_price: token_b_price as u64 };
+            let pool_token_amount = pool_token_amount as u128;
+            let pool_token_supply = pool_token_supply as u128;
+            let swap_token_a_amount = swap_token_a_amount as u128;
+            let swap_token_b_amount = swap_token_b_amount as u128;
+            let token_b_price = token_b_price as u128;
+
+            let value = curve.normalized_value(swap_token_a_amount, swap_token_b_amount).unwrap();
+
+            // Make sure we trade at least one of each token
+            prop_assume!(pool_token_amount * value >= 2 * token_b_price * pool_token_supply);
+            prop_assume!(pool_token_amount <= pool_token_supply);
+            let withdraw_result = curve
+                .pool_tokens_to_trading_tokens(
+                    pool_token_amount,
+                    pool_token_supply,
+                    swap_token_a_amount,
+                    swap_token_b_amount,
+                    RoundDirection::Floor,
+                )
+                .unwrap();
+            prop_assume!(withdraw_result.token_a_amount <= swap_token_a_amount);
+            prop_assume!(withdraw_result.token_b_amount <= swap_token_b_amount);
```

Comment:
Should these be asserts?

Reply:
This was a bit of a hammer solution to make sure it's a working withdrawal.  When withdrawing equally on both sides, constant price always uses the same ratio, so you can easily get into a situation where you try to overdraw one of the sides.  For example, if you have a price of 20 per token b, 1000 token A, and 50 token B, you'll never run into any issues since everything is in perfect balance.  If you try to withdraw 10% of pool tokens, you'll try to take 10% of the value of the pool in equal A and B, meaning 100 A and 5 B.  Now if your supplies are 10000 token A, and 5 token B, and you try to withdraw 10% of the value of the pool, equally split, you'll ask for 505 token A and 25 token B.  It's one of the big reasons we needed single-sided deposit and withdrawal.  Other than by doing the same math as the withdrawal, I didn't know how to make sure the withdrawal was valid.

Reply:
I see, thanks for the explanation. The way the test is written seems fine to me